// xero-node/dist/gen/model/payroll-au/earningsType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var EarningsType;
(function (EarningsType) {
    EarningsType[EarningsType["FIXED"] = 'FIXED'] = "FIXED";
    EarningsType[EarningsType["ORDINARYTIMEEARNINGS"] = 'ORDINARYTIMEEARNINGS'] = "ORDINARYTIMEEARNINGS";
    EarningsType[EarningsType["OVERTIMEEARNINGS"] = 'OVERTIMEEARNINGS'] = "OVERTIMEEARNINGS";
    EarningsType[EarningsType["ALLOWANCE"] = 'ALLOWANCE'] = "ALLOWANCE";
    EarningsType[EarningsType["LUMPSUMD"] = 'LUMPSUMD'] = "LUMPSUMD";
    EarningsType[EarningsType["EMPLOYMENTTERMINATIONPAYMENT"] = 'EMPLOYMENTTERMINATIONPAYMENT'] = "EMPLOYMENTTERMINATIONPAYMENT";
    EarningsType[EarningsType["LUMPSUMA"] = 'LUMPSUMA'] = "LUMPSUMA";
    EarningsType[EarningsType["LUMPSUMB"] = 'LUMPSUMB'] = "LUMPSUMB";
    EarningsType[EarningsType["BONUSESANDCOMMISSIONS"] = 'BONUSESANDCOMMISSIONS'] = "BONUSESANDCOMMISSIONS";
    EarningsType[EarningsType["LUMPSUME"] = 'LUMPSUME'] = "LUMPSUME";
})(EarningsType = exports.EarningsType || (exports.EarningsType = {}));
